Feat modular arithmetic

welkerje opened the pull request:
## Modular Arithmetic Generators

### generators
added multiple variants for questions on the topic of modular arithmetic
- simple calculations
- reduction of large^{TM} numbers
- find modular inverse
- modular exponentiation
- Chinese Remainder Theorem

### utilies
added some math utilities
- gcd
- calculateModularInverse
- modular exponentiation
- solveCRT
- areCoprime

### ideas
- (Extended) Euclidean Algorithm
- RSA

AbcAeffchen reviewed (commented):
The PR looks good. I like the named placeholders in the translation texts.

The only thing I find odd is that some files look like prettier should complain about the indends (4 spaces instead of 2) and the semicolons at the end of statements. For some reasons they pass the automatic tests. However you should run `bun format` before we merge the branch.

Fabianstw commented on shared/src/question-generators/math/modTricks.ts:
```diff
@@ -0,0 +1,292 @@
+        }
+        return (userAnswer - a) % b === 0
+            ? { correct: true, feedbackText: t(translations, lang, "feedbackCorrect") }
+            : { correct: false, feedbackText: t(translations, lang, "feedbackIncorrect") };
```
Maybe add: `correctAnswer: a.toString()` or `...(a * random.int(1, 5).toString())`, such that the user can see a correct solution, but maybe only `a` is to obvious

Fabianstw commented on shared/src/question-generators/math/modTricks.ts:
```diff
@@ -0,0 +1,292 @@
+
+// Reduction
+function generateReductionQuestion(lang: Language, path: string, random: Random) {
+    const x = random.int(100, 999);
```
Maybe allowing negative numbers too?

AbcAeffchen replied:
Then one needs to specify what reduction is meant. Maybe this is needed anyway. A reduction `x mod n` could mean to write x in an interval [0,n-1] or [-n/2+1,n/2].

Fabianstw commented on shared/src/question-generators/math/modTricks.ts:
```diff
@@ -0,0 +1,292 @@
+        }
+        return userAnswer === correctAnswer
+            ? { correct: true, feedbackText: t(translations, lang, "feedbackCorrect") }
+            : { correct: false, feedbackText: t(translations, lang, "feedbackIncorrect") };
```
Correctanswer for the user like above 👍

Fabianstw commented on shared/src/question-generators/math/modTricks.ts:
```diff
@@ -0,0 +1,292 @@
+    return { question, testing };
+}
+
+function getCRTFeedbackFunction(lang: Language, crtValue: number, commonModulus: number): FreeTextFeedbackFunction {
```
Maybe also a check format function? So the user can see if the input has the correct type

Fabianstw commented on shared/src/question-generators/math/modTricks.ts:
```diff
@@ -0,0 +1,292 @@
+
+        return userModulus !== commonModulus || userValue !== crtValue
+            ? { correct: false, feedbackText: t(translations, lang, "feedbackIncorrect") }
+            : { correct: true, feedbackText: t(translations, lang, "feedbackCorrect") };
```
Here probably a correctAnswer too.

Fabianstw commented on shared/src/question-generators/math/modTricks.ts:
```diff
@@ -0,0 +1,292 @@
+        {
+            type: "string",
+            name: "variant",
+            allowedValues: ["simple", "reduction", "inverse", "exponentiation", "crt"],
```
The variant `crt` is a little more complex / different to the other ones. Maybe we can separate this into an own generator with variants for different difficulties, because congruent system of 4 (for me) is quite difficult. Just an idea

AbcAeffchen replied:
Yes, would make sense.

Fabianstw commented on shared/src/question-generators/math/modularArithmetic.test.ts:
```diff
@@ -0,0 +1,128 @@
+import { describe, expect, test } from "vitest"
+import { FreeTextFeedback, FreeTextQuestion } from "../../api/QuestionGenerator"
+import { ModularArithmetic } from "./modularArithmetic"
+import { sampleRandomSeed } from "../../utils/random"
```
Please use relative imports :-)

Fabianstw reviewed (commented):
Looks very good for me! I like the questions about this topic :-) 

Just a few short comments 👍

AbcAeffchen commented on shared/src/question-generators/math/modTricks.ts:
```diff
@@ -0,0 +1,292 @@
+        reductionQuestion: "Reduce ${{x}}$ modulo ${{n}}$.",
+        inverseQuestion: "Find the modular inverse of ${{a}}$ modulo ${{n}}$.",
+        expQuestion: "Calculate ${{a}}^{{{b}}} \\pmod{ {{n}} }$.",
+        simpleQuestion: "Find an integer $x$ such that $x ≡ {{a}} \\pmod{ {{b}} }$.",
```
I don't know if latex works with such unicode symbols. So to be safe, using `\equiv` would probably be better.

welkerje commented:
I implemented most of the suggested changes.
Since I wasn't happy with the reduction question anyway,  I replaced it with some new generators that serve the intention better.

AbcAeffchen commented:
I had a look at the CRT questions and found this one:
tcs.uni-frankfurt.de/algo-learn-testing/pull_request_40/en/crt/3/mgir3wk

Is it intended that the question contains "x \equiv 3 (mod 3)"? I would have expected this to be written as "x \equiv 0 (mod 3)"

In the Modular Arithmetic with Factorization question, the multiplication variant uses `\times` instead of `\cdot`. Also, a lot of the solutions were zero. Not sure if this happens unproportionally often or if I just had bad luck.

I also encountered a strange message on this questions:
tcs.uni-frankfurt.de/algo-learn-testing/pull_request_40/en/modtricks/fermatExponentiation/mq30ot0
When entering a false answer I get: "The correct answer is: 3.59\eqiv9(mod10)." Here seems the latex to be broken. And for some reason the answer is not an integer.

welkerje commented:
I assume the 0s show up when the modulus is small? I think that would that is due to small moduli having a high likelihood of being a prime factor of one of the generated factors. This could, for example, be avoided by setting the lower bound of allowed moduli higher (currently 2).

AbcAeffchen commented:
I see. Maybe it is an option to randomly draw from a set of prim numbers and compute the values shown in the task from them?

holgerdell commented:
Or first uniformly sample the desired result (the modulus), and then sample a random question that has this result as its solution. In this way, all solutions would be equally likely

Fabianstw commented on shared/src/question-generators/math/crt.ts:
```diff
@@ -0,0 +1,166 @@
+        ? t(translations, lang, "feedbackCorrect")
+        : t(translations, lang, "feedbackIncorrect", {
+            correctAnswer: `${crtValue} (mod ${commonModulus})`,
+          }),
```
Case the answer is correct, I would only return `{correct: isCorrect}`
Otherwise: 
```
return {
      correct: isCorrect,
      correctAnswer: isCorrect ? "" : `${crtValue} (mod ${commonModulus})`
}
```
Something like this, because we already have a block for the correctAnswer and feedbackText: 
tcs.uni-frankfurt.de/algo-learn/en/plnf/input/2/2b7t26o

This is also for some other questions

Fabianstw commented on shared/src/question-generators/math/utils.ts:
```diff
@@ -0,0 +1,107 @@
+  maxFactors: number = 3,
+  minValue: number = 2,
+  maxValue: number = 10,
+): number[] {
```
Maybe the params as dict, such that I could use standards for min/maxFactors and only provide own values for min/maxValue

Fabianstw commented on shared/src/question-generators/math/modFactorization.ts:
```diff
@@ -0,0 +1,217 @@
+    }
+  } while ((a * b) % n !== result && attempts < maxAttempts + 1000)
+  if ((a * b) % n !== result) {
+    throw new Error(`Timeout at factor generation after ${maxAttempts + 1000} attempts`)
```
How likely is this error? Or instead of throwing an error, can we maybe change the calculation such that we always generate a question. 
But if the error is very unlikely just ignore :-)

Fabianstw commented on shared/src/question-generators/math/modTricks.ts:
```diff
@@ -0,0 +1,348 @@
+    return {
+      question,
+      testing,
+    }
```
Maybe directly return the result/testing. Like in the other generators. Would be simpler to read

Fabianstw reviewed (commented):
Only some minor comments and missing docStrings over a couple of functions. 

And please replace `\\times` with `\\cdot` :-)
